code/modules/power/tesla/energy_ball.dm

```diff
@@ -18,93 +30,118 @@
+	
 	dissipate = TRUE //Do we lose energy over time?
-	dissipate_delay = 5
-	time_since_last_dissipiation = 0
-	dissipate_strength = 1
+	dissipate_delay = 10 SECONDS
+
 	var/produced_power
 	var/energy_to_raise = 32
 	var/energy_to_lower = -20
+	var/list/shocked_things = list()
 
 /obj/anomaly/energy_ball/Initialize(mapload, starting_energy = 50, is_miniball = FALSE)
 	. = ..()
+
 	energy = starting_energy
+	miniball = is_miniball
 	START_PROCESSING(SSobj, src)
-	var/turf/spawned_turf = get_turf(src)
-	message_admins("A tesla has been created at [ADMIN_VERBOSEJMP(spawned_turf)].")
-	investigate_log("(tesla) was created at [AREACOORD(spawned_turf)].", INVESTIGATE_ENGINES)
 
-/obj/anomaly/energy_ball/ex_act(severity, target)
-	return
+	if (!is_miniball)
+		set_light(10, 7, "#5e5edd")
+
+		var/turf/spawned_turf = get_turf(src)
+		message_admins("A tesla has been created at [ADMIN_VERBOSEJMP(spawned_turf)].")
+		investigate_log("was created at [AREACOORD(spawned_turf)].", INVESTIGATE_ENGINE)
 
 /obj/anomaly/energy_ball/Destroy()
+	if(orbiting && istype(orbiting.parent, /obj/anomaly/energy_ball))
+		var/obj/anomaly/energy_ball/parent_energy_ball = orbiting.parent
+		parent_energy_ball.orbiting_balls -= src
+
 	QDEL_LIST(orbiting_balls)
 	STOP_PROCESSING(SSobj, src)
-	. = ..()
+
+	return ..()
+
+/obj/anomaly/energy_ball/ex_act(severity, target)
+	return
 
 /obj/anomaly/energy_ball/process(delta_time)
-	handle_energy(delta_time)
+	if(orbiting)
+		energy = 0 // ensure we dont have miniballs of miniballs
+	else
+		handle_energy(delta_time)
 
-	move_the_basket_ball(4 + orbiting_balls.len * 1.5)
+	if(last_move <= world.time + 4 SECONDS)
+		last_move = world.time
+		move(4 + (orbiting_balls.len * 1.5 ))
 
-	playsound(src.loc, 'sound/magic/lightningbolt.ogg', 100, 1, extrarange = 30)
+		playsound(src.loc, 'sound/magic/lightningbolt.ogg', 70, TRUE, extrarange = 30)
 
-	pixel_x = 0
-	pixel_y = 0
+		pixel_x = 0
+		pixel_y = 0
+		shocked_things.Cut(1, shocked_things.len / 1.3)
+		var/list/shocking_info = list()
+		tesla_zap(src, 3, TESLA_DEFAULT_POWER, shocked_targets = shocking_info)
 
-	//Main one can zap
-	//Tesla only zaps if the tick usage isn't over the limit.
-	if(!TICK_CHECK)
-		tesla_zap(src, 7, TESLA_DEFAULT_POWER, TESLA_ENERGY_PRIMARY_BALL_FLAGS)
-	else
-		//Weaker, less intensive zap
-		tesla_zap(src, 4, TESLA_DEFAULT_POWER, TESLA_ENERGY_MINI_BALL_FLAGS)
 		pixel_x = -32
 		pixel_y = -32
-		return
-
-	pixel_x = -32
-	pixel_y = -32
-	for (var/ball in orbiting_balls)
-		if(TICK_CHECK)
-			return
-		var/range = rand(1, CLAMP(orbiting_balls.len, 3, 7))
-		//Miniballs don't explode.
-		tesla_zap(ball, range, TESLA_MINI_POWER/7*range, TESLA_ENERGY_MINI_BALL_FLAGS)
+		for (var/ball in orbiting_balls)
+			var/range = rand(1, clamp(orbiting_balls.len, 2, 3))
+			var/list/temp_shock = list()
+			//We zap off the main ball instead of ourselves to make things looks proper
+			tesla_zap(src, range, TESLA_MINI_POWER/7*range, shocked_targets = temp_shock)
```

Comment:
Won't this re-introduce the bug that maps the miniball zaps cause explosions, which results in absolute carnage and lag?

Reply:
Hm. Yes, I see. I think I have a solution for that, one sec